pageserver/src/page_service.rs

```diff
@@ -932,8 +933,33 @@ impl postgres_backend_async::Handler for PageServerHandler {
     }
 }
 
-fn get_local_timeline(tenant_id: TenantId, timeline_id: TimelineId) -> Result<Arc<Timeline>> {
-    tenant_mgr::get_tenant(tenant_id, true)
+/// Get active tenant.
+///
+/// If the tenant is Loading, waits for it to become Active, for up to 30 s. That
+/// ensures that queries don't fail immediately after pageserver startup, because
+/// all tenants are still loading.
+async fn get_active_tenant_with_timeout(tenant_id: TenantId) -> Result<Arc<Tenant>> {
+    let tenant = tenant_mgr::get_tenant(tenant_id, false)?;
+
+    match tokio::time::timeout(Duration::from_secs(30), tenant.wait_until_loaded()).await {
+        Ok(result) => {
+            let state = result?;
+            if !matches!(state, pageserver_api::models::TenantState::Active { .. }) {
```

Comment:
I think instead of `wait_until_loaded`, we should have a `tenant.wait_until_active` function, pushing this `matches!(...)` here down into the `tenant.rs`.

There's no other user of `tenant.wait_until_loaded()`, and strictly speaking, it doesn't wait until we reach some 'loaded' state, but it waits until we're in some state other than `Loaded`.

Reply:
Hm, also, what about state `Attaching` ? I think in this file, we shouldn't care about the source state (`Loading` or `Attaching`, only the desired target state `Active`).

Will code something up.